src/filters/list.js

```diff
@@ -0,0 +1,196 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module pastefromoffice/filters/list
+ */
+
+import Element from '@ckeditor/ckeditor5-engine/src/view/element';
+import Matcher from '@ckeditor/ckeditor5-engine/src/view/matcher';
+import Position from '@ckeditor/ckeditor5-engine/src/view/position';
+import TreeWalker from '@ckeditor/ckeditor5-engine/src/view/treewalker';
+import UpcastWriter from '@ckeditor/ckeditor5-engine/src/view/upcastwriter';
+
+/**
+ * Transforms Word specific list-like elements to the semantic HTML lists.
+ *
+ * Lists in Word are represented by block elements with special attributes like:
+ *
+ *		<p class=MsoListParagraphCxSpFirst style='mso-list:l1 level1 lfo1'>...</p> // Paragraph based list.
+ *		<h1 style='mso-list:l0 level1 lfo1'>...</h1> // Heading 1 based list.
+ *
+ * @param {module:engine/view/node~Node|module:engine/view/documentfragment~DocumentFragment} bodyView The view
+ * structure which to transform.
+ * @returns {module:engine/view/node~Node|module:engine/view/documentfragment~DocumentFragment} The view
+ * structure instance with list-like elements transformed into semantic lists.
+ */
+export function paragraphsToLists( bodyView, stylesString ) {
+	const firstChild = bodyView.getChild( 0 );
+
+	if ( firstChild ) {
+		const listNodes = findAllListNodes( Position.createBefore( firstChild ) );
+		createLists( listNodes, stylesString );
+	}
+
+	return bodyView;
+}
+
+// Writer used for View elements manipulation.
+const writer = new UpcastWriter();
+
+// Matcher for finding list-like elements.
+const listMatcher = new Matcher( {
+	name: /^p|h\d+$/,
+	styles: {
+		'mso-list': /.*/
+	}
+} );
+
+// Matcher for finding `span` elements holding lists numbering/bullets.
+const listBulletMatcher = new Matcher( {
+	name: 'span',
+	styles: {
+		'mso-list': 'Ignore'
+	}
+} );
+
+// Finds all list-like nodes starting from a given position.
+//
+// @param {module:engine/src/view/position~Position} startPosition Position from which to start looking.
+// @returns {Array.<Object>} Array of found list items. Each item is an object containing:
+//
+//		* {module:engine/src/view/element~Element} element List-like element.
+//		* {Number} id List item id parsed from `mso-list` style (see `getListItemData()` function).
+//		* {Number} order List item creation order parsed from `mso-list` style (see `getListItemData()` function).
+//		* {Number} indent List item indentation level parsed from `mso-list` style (see `getListItemData()` function).
+function findAllListNodes( startPosition ) {
+	const treeWalker = new TreeWalker( { startPosition, ignoreElementEnd: true } );
+
+	// Find all list nodes.
+	const listNodes = [];
```

Comment:
Blank line missing (yep, there's a ticket to improve ESLint's config but there was some kind of a problem).